docs/TypedMongo.md

```diff
@@ -0,0 +1,447 @@
+an _implied filter_ that is automatically included into the query if this reference is used in a MongoDB query document
+or projection.
+
+`MongoPropertyRef` is a fundamental type that serves as an entry point for creating queries, projections, updates,
+sort orders, indices, etc.
+
+### `MongoDocumentFilter`
+
+A `MongoDocumentFilter[E]` represents a MongoDB query document for an entity type `E`. Usually, filters are created
+via `MongoPropertyRef`s, e.g. 
+
+```scala
+val query: MongoDocumentFilter[MyEntity] = MyEntity.ref(_.int) > 10 && MyEntity.ref(_.data.flag).is(true)
+```
+
+For more examples, see the Scaladoc or 
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/MongoFilterTest.scala).
+
+### `MongoProjection`
+
+A `MongoProjection[E]` represents a MongoDB projection document that specifies the properties that should be included
+in the response of a query. It is usually one of:
+
+* The whole-document projection, e.g. `MyEntity.SelfRef`.
+* The whole-document projection narrowed to a single case class or intermediate subtrait in a sealed hierarchy, e.g.
+  `MyUnionEntity.as[FirstCase]`.
+* A single `MongoPropertyRef`, e.g. `MyEntity.ref(_.int)`
+* An arbitrary combination of projections into a tuple, e.g. `MongoProjection.zip(MyEntity.ref(_.int), MyEntity.ref(_.data))`
+
+For more examples, see the Scaladoc or
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/MongoProjectionTest.scala).
+
+### `MongoDocumentOrder`
+
+A `MongoDocumentOrder[E]` represents a sort order for a document of type `E`. It may be one of:
+
+* Unspecified (empty) order, i.e. `MongoDocumentOrder.unspecified`
+* Single-field order, e.g. `MyEntity.ref(_.int).ascending`
+* Multi-field order, e.g. `MongoDocumentOrder(MyEntity.ref(_.int) -> true, MyEntity.ref(_.data.number) -> false)`
+
+For more examples, see the Scaladoc or
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/MongoOrderTest.scala).
+
+### `MongoDocumentUpdate`
+
+A `MongoDocumentUpdate[E]` represents a MongoDB update document for an entity of type `E`. Example:
+
+```scala
+val update: MongoDocumentUpdate[MyEntity] = 
+  MyEntity.ref(_.strOpt).set(Opt("foo")) && MyEntity.ref(_.int).inc(5)
+```
+
+For more examples, see the Scaladoc or
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/MongoUpdateTest.scala).
+
+### `MongoIndex`
+
+A `MongoIndex[E]` represents a MongoDB index document for an entity of type `E`. Examples:
+
+* Single field index, e.g. `MyEntity.ref(_.int).ascendingIndex`
+* Multi-field index, e.g.
+
+  ```scala
+  import MongoIndexType._
+  val index: MongoIndex[MyEntity] =
+  MongoIndex(MyEntity.ref(_.str) -> Hashed, MyEntity.ref(_.int) -> Descending)
```

Comment:
This would actually be an error (hashed indexes cannot be part of a compound index). Maybe the API is too liberal?

Reply:
Yeah I should rethink `MongoIndex`

Reply:
Well that's actually not true anymore: https://docs.mongodb.com/manual/core/index-hashed/#create-a-compound-hashed-index

Reply:
I know that this API still allows creation of many invalid indexes but I think there's no point in trying to cover this with very precise static API - it will likely be complicated and become incompatible as MongoDB is evolving.

Reply:
Whoa that's fresh. I'm okay with this being a bit liberal then, but I'm still not the biggest fan of this API in the sense that single indexes are created via extension methods on refs and compound indexes have a completely different feel (making them harder to discover I guess?). Maybe that's a good thing though given how different such indexes are.